salt/utils/json.py

```diff
@@ -26,7 +26,12 @@ def find_json(raw):
     '''
     ret = {}
     for ind, _ in enumerate(raw):
```

Comment:
I want to say more here.
This `for` loops iterates the `raw` that is a string according the docstring. So `ind` here is a string character number. But below it's used as a line number in `raw` string data.
So as for me overall logic in this function should look like this:
```python
def find_json(raw):
    '''
    Pass in a raw string and load the json when it starts. This allows for a
    string to start with garbage and end with json but be cleanly loaded
    '''
    ret = {}
    for ind, _ in enumerate(raw):
        working = raw[ind:]
        try:
            ret = json.loads(raw[ind:])  # future lint: blacklisted-function
        except ValueError:
            continue
        if ret:
            return ret
    if not ret:
        # Not json, raise an error
        raise ValueError
```

Reply:
I talked with @thatch45 and confirmed the code should try to load json one line at a time, not one character at a time. This is the reason for splitting and joining based on a line index. Using the `enumerate()` on the entire string appears to have been a simple oversight. In reality, this would have resulted in the for loop doing a _lot_ of extra `json.loads()` calls on empty strings, since slicing returns an empty list when the slice indices are outside of the actual size of the list:

```python
>>> x = ['foo', 'bar']
>>> x[5:]
[]
```

Joining an empty list like this would produce an empty string, and loading an empty string using `json.loads()` raises a `ValueError`, which we are catching (hence this not being caught before).

So, a more correct change would be:

```python
ret = {}
lines = raw.splitlines()
for ind, _ in enumerate(lines):
    try:
        working = '\n'.join(lines[ind:])
    except UnicodeDecodeError:
        working = '\n'.join(salt.utils.data.decode(lines[ind:]))

    try:
        ret = json.loads(working)  # future lint: blacklisted-function
    except ValueError:
        continue
    if ret:
        return ret
if not ret:
    # Not json, raise an error
    raise ValueError
```

Reply:
Working on this fix now, as well as a unit test.